src/parser/postgresparser.cpp

```diff
@@ -2026,4 +2034,71 @@ std::unique_ptr<VariableSetStatement> PostgresParser::VariableSetTransform(Parse
   return result;
 }
 
+// Postgres.SelectStmt.withClause -> terrier.TableRef
+std::vector<std::unique_ptr<TableRef>> PostgresParser::WithTransform(ParseResult *parse_result, WithClause *root) {
+  // Postgres parses 'SELECT;' to nullptr
+  std::vector<std::unique_ptr<TableRef>> ctes;
+  if (root == nullptr) {
+    return ctes;
+  }
+
+  // TODO(Rohan, Preetansh, Gautam): - HANDLE CASE WHEN LENGTH OF ROOT > 1
+  std::unique_ptr<TableRef> result = nullptr;
+  ListCell *current = root->ctes_->head;
+  while (current != nullptr) {
+    auto node = reinterpret_cast<Node *>(current->data.ptr_value);
+    auto common_table_expr = reinterpret_cast<CommonTableExpr *>(node);
+    auto cte_query = reinterpret_cast<Node *>(common_table_expr->ctequery_);
+    switch (cte_query->type) {
+      case T_SelectStmt: {
+        auto cte_select_query = reinterpret_cast<SelectStmt *>(cte_query);
+        if (root->iterative_ || root->recursive_) {
+          auto tmp = cte_select_query->larg_;
+          cte_select_query->larg_ = cte_select_query->rarg_;
+          cte_select_query->rarg_ = tmp;
+        }
+        auto select = SelectTransform(parse_result, cte_select_query);
+        if (select == nullptr) {
+          current = current->next;
+          continue;
+        }
+        auto alias = common_table_expr->ctename_;
+
+        std::vector<parser::AliasType> colnames;
+        auto col_names_root = common_table_expr->aliascolnames_;
+        if (col_names_root != nullptr) {
+          size_t i = 0;
+          for (auto cell = col_names_root->head; cell != nullptr; cell = cell->next) {
+            auto target = reinterpret_cast<Value *>(cell->data.ptr_value);
+            auto column = target->val_.str_;
+            colnames.emplace_back(parser::AliasType(column, i));
+            i++;
+          }
+        }
+        //        if (colnames.empty()) {
+        //          for (auto &expr : select->GetSelectColumns()) {
+        //            // TODO(tanujnay112) route the "?column?" codepath through traffic cop
+        //            colnames.push_back(expr->GetAlias().empty() ? "?column?" : expr->GetAlias());
+        //          }
+        //        }
```

Comment:
CTEs are a little complicated in how column names get picked -- if you consider the query

`WITH xxx(a) AS (SELECT 1, b as c, d, 2) ...`

The table `xxx` will have four columns: 

1. The first column name is `a`, since this was specified with the CTE name in `WITH xxx(a)`
2. The second column name is `c`, since no name was specified for the second column, and the alias for the second column is `c`
3. The third column name is `d`, since neither of the above two items were specified, so the column name itself is used
4. The fourth column name is `?column?`, since there is no column name

We do all of this casing in the binder, once column names/`*`'s have been bound to specific columns, so we handle the fourth case (where none of the other three options provide a column name to use) in the binder as well.